coreneuron/io/phase2.cpp

```diff
@@ -0,0 +1,1211 @@
+            }
+        }
+    }
+
+    if (nrn_have_gaps && interleave_permute_type) {
+        nrn_partrans::gap_indices_permute(nt);
+    }
+
+    /* here we setup the mechanism dependencies. if there is a mechanism dependency
+     * then we allocate an array for tml->dependencies otherwise set it to nullptr.
+     * In order to find out the "real" dependencies i.e. dependent mechanism
+     * exist at the same compartment, we compare the nodeindices of mechanisms
+     * returned by nrn_mech_depend.
+     */
+
+    /* temporary array for dependencies */
+    int* mech_deps = (int*)ecalloc(memb_func.size(), sizeof(int));
+
+    for (auto tml = nt.tml; tml; tml = tml->next) {
+        /* initialize to null */
+        tml->dependencies = nullptr;
+        tml->ndependencies = 0;
+
+        /* get dependencies from the models */
+        int deps_cnt = nrn_mech_depend(tml->index, mech_deps);
+
+        /* if dependencies, setup dependency array */
+        if (deps_cnt) {
+            /* store "real" dependencies in the vector */
+            std::vector<int> actual_mech_deps;
+
+            Memb_list* ml = tml->ml;
+            int* nodeindices = ml->nodeindices;
+
+            /* iterate over dependencies */
+            for (int j = 0; j < deps_cnt; j++) {
+                /* memb_list of dependency mechanism */
+                Memb_list* dml = nt._ml_list[mech_deps[j]];
+
+                /* dependency mechanism may not exist in the model */
+                if (!dml)
+                    continue;
+
+                /* take nodeindices for comparison */
+                int* dnodeindices = dml->nodeindices;
+
+                /* set_intersection function needs temp vector to push the common values */
+                std::vector<int> node_intersection;
+
+                /* make sure they have non-zero nodes and find their intersection */
+                if ((ml->nodecount > 0) && (dml->nodecount > 0)) {
+                    std::set_intersection(nodeindices, nodeindices + ml->nodecount, dnodeindices,
+                                          dnodeindices + dml->nodecount,
+                                          std::back_inserter(node_intersection));
+                }
+
+                /* if they intersect in the nodeindices, it's real dependency */
+                if (!node_intersection.empty()) {
+                    actual_mech_deps.push_back(mech_deps[j]);
+                }
+            }
+
+            /* copy actual_mech_deps to dependencies */
+            if (!actual_mech_deps.empty()) {
+                tml->ndependencies = actual_mech_deps.size();
+                tml->dependencies = (int*)ecalloc(actual_mech_deps.size(), sizeof(int));
+                memcpy(tml->dependencies, &actual_mech_deps[0],
+                       sizeof(int) * actual_mech_deps.size());
+            }
+        }
+    }
+
+    /* free temp dependency array */
+    free(mech_deps);
+
+    /// Fill the BA lists
```

Comment:
In fact, for this part and many before.. these should be split into smaller private functions here. This one should be `fill_ba_lists()` etc.

Reply:
I have made 3 more

Reply:
I will continue later